github/commands/pull_request.py

```diff
@@ -74,59 +77,95 @@ def on_select_action(self, idx):
 
-        if branch_name:
-            self.window.status_message("Creating local branch for PR...")
-            self.git(
-                "branch",
-                branch_name,
-                self.pr["head"]["sha"]
-            )
+        clone_url = self.pr["head"]["repo"]["clone_url"]
+        ssh_url = self.pr["head"]["repo"]["ssh_url"]
+        ref = self.pr["head"]["ref"]
 
-        self.window.status_message("Checking out PR...")
-        self.checkout_ref(branch_name or self.pr["head"]["sha"])
+        url = ssh_url if self.base_remote_url.startswith("git@") else clone_url
+        self.git("fetch", url, ref)
+        self.checkout_ref(ref)
         util.view.refresh_gitsavvy_interfaces(self.window, refresh_sidebar=True)
 
-    def create_branch_for_pr(self, branch_name):
-        self.window.status_message("Fetching PR commit...")
-        self.git(
-            "fetch",
-            self.pr["head"]["repo"]["clone_url"],
-            self.pr["head"]["ref"]
-        )
+    def create_branch_for_pr(self, branch_name, checkout=False, ask_set_upstream=True):
+        if not branch_name:
+            return
 
         self.window.status_message("Creating local branch for PR...")
-        self.git(
-            "branch",
-            branch_name,
-            self.pr["head"]["sha"]
-        )
+        set_upstream = False
+
+        owner = self.pr["head"]["repo"]["owner"]["login"]
+        if owner == self.base_remote.owner:
+            owner = self.base_remote_name
+            # set upstream automatically if the pr is from integrated remote
+            ask_set_upstream = False
+            set_upstream = True
+
+        clone_url = self.pr["head"]["repo"]["clone_url"]
+        ssh_url = self.pr["head"]["repo"]["ssh_url"]
+        ref = self.pr["head"]["ref"]
+
+        if ask_set_upstream:
+            set_upstream = sublime.ok_cancel_dialog(
+                "Set upstream to '{}/{}'?".format(owner, ref))
+
+        if set_upstream:
+            if owner not in self.remotes.keys():
+                url = ssh_url if self.base_remote_url.startswith("git@") else clone_url
+                self.git("remote", "add", owner, url)
+
+            self.create_branch_from_remote_for_pr(branch_name, owner, ref, checkout)
+        else:
+            self.create_branch_from_sha_for_pr(branch_name, owner, checkout)
+
+    def create_branch_from_remote_for_pr(self, branch_name, remote, remote_branch, checkout):
+        self.git("fetch", remote, remote_branch)
+        ref = "{}/{}".format(remote, remote_branch)
+        self.git("branch", branch_name, ref)
+        self.git("branch", "-u", ref, branch_name)
+
+        if checkout:
+            self.checkout_ref(branch_name)
+
+        util.view.refresh_gitsavvy_interfaces(self.window, refresh_sidebar=True)
+
+    def create_branch_from_sha_for_pr(self, branch_name, remote, checkout):
+        self.git("fetch", remote, self.pr["head"]["sha"])
```

Comment:
Honestly, I feel stupid, but for this branch of the code, t.i. we don't set upstream, we can't assume we have a remote set. Can we? We need to fetch from the URL I would have thought.

Also why Sha and not ref?

Reply:
Oh, you are right.